compiler/src/dotty/tools/dotc/config/PathResolver.scala

```diff
@@ -36,17 +36,25 @@ object PathResolver {
   /** Values found solely by inspecting environment or property variables.
    */
   object Environment {
-    private def searchForBootClasspath = (
-      systemProperties find (_._1 endsWith ".boot.class.path") map (_._2) getOrElse ""
-    )
+    private def searchForBootClasspath = {
+      import scala.jdk.CollectionConverters.*
+      val props = System.getProperties
+      // This formulation should be immune to ConcurrentModificationExceptions when system properties
+      // we're unlucky enough to witness a partially published result of System.setProperty or direct
+      // mutation of the System property map. stringPropertyNames internally uses the Enumeration interface,
+      // rather than Iterator, and this disables the fail-fast ConcurrentModificationException.
+      val propNames = props.stringPropertyNames()
+      propNames.asScala collectFirst { case k if k endsWith ".boot.class.path" => props.getProperty(k) } getOrElse ""
+    }
```

Comment:
I tried to understand why this fallback exists, but I landed [here](https://github.com/scala/scala/commit/171d21f11a58e905de8f96a9bac41f6aced3e384#diff-27814b1c8a7e1970b14494224ad79a879446d59765649f92fd4e05b0b81e2dadR44) and that was a dead end. Maybe we could just remove this code and do 

```scala
def javaBootClassPath: String = propOrElse("sun.boot.class.path", "")
```

Reply:
probably, but for now let's go with that tried-and-true

Reply:
I see that this is what I asked in my other comment: it is because scala 2 runner supplied `scala.boot.class.path`. It would be worth revisiting what is supported under scala-cli. There is still `-bootclasspath` option. In scala 2, `-nobootcp` meant don't use `-Xbootclasspath` and don't supply `scala.boot.classpath`. Modern class loader hierarchy is quite different. I always use `-nobootcp` in scala 2.